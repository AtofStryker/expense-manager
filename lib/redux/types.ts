import { State } from '../state'
import { Action as ReduxAction } from 'redux'
import { ThunkAction } from 'redux-thunk'
import { Logger } from '../types'
import { Api } from '../api'

export interface ThunkExtraArgument {
  logger: Logger
  api: Api
}

export interface Action<Payload = void> extends ReduxAction<string> {
  loggable?: boolean
  payload?: Payload
  reducer: (state: State) => State
}

export type Thunk<Result = Promise<unknown>> = ThunkAction<
  Result,
  State,
  ThunkExtraArgument,
  Action<unknown>
>
