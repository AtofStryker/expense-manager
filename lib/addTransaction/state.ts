import { DEFAULT_CURRENCY } from '../shared/currencies'
import { ObjectOf } from '../types'

export enum TransactionTypes {
  fromUser = 'fromUser',
  imported = 'imported',
  repeated = 'repeated',
}

export type TransactionType = keyof typeof TransactionTypes

export interface Tag {
  id: string
  name: string
  uid: string
}

export enum RepeatingOptions {
  none = 'none',
  inactive = 'inactive',
  daily = 'daily',
  weekly = 'weekly',
  monthly = 'monthly',
  annually = 'annually',
}

export type RepeatingOption = keyof typeof RepeatingOptions

export interface BaseTransaction {
  transactionType: TransactionType
  // NOTE: order might be important
  tagIds: string[]
  currency: string
  isExpense: boolean
  note: string
  repeating: RepeatingOption
}

export interface Transaction extends BaseTransaction {
  id: string
  dateTime: Date
  amount: number
  uid: string
}

export interface AddTransaction extends BaseTransaction {
  useCurrentTime: boolean
  newTags: ObjectOf<Tag>
  dateTime?: Date
  tagInputValue: string
  amount: string
  shouldValidateAmount: boolean
}

export const createDefaultAddTransactionState = (): AddTransaction => ({
  transactionType: TransactionTypes.fromUser,
  amount: '',
  tagIds: [],
  newTags: {},
  currency: DEFAULT_CURRENCY.value,
  tagInputValue: '',
  isExpense: true,
  note: '',
  dateTime: undefined,
  useCurrentTime: true,
  shouldValidateAmount: false,
  repeating: RepeatingOptions.none,
})
