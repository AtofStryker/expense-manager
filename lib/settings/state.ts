import { DEFAULT_CURRENCY, Currency } from '../shared/currencies'
import { FirebaseField } from '../types'

export interface Settings {
  mainCurrency: Currency
  defaultCurrency: Currency
}
export interface ExchangeRates {
  rates: { [key in Currency]: number }
  base: Currency
  date: string
}

export interface Profile extends FirebaseField {
  settings: Settings
  exchangeRates: ExchangeRates
}

export const getInitialState = (user: firebase.User): Profile => ({
  uid: user.uid,
  id: user.uid,
  settings: {
    mainCurrency: DEFAULT_CURRENCY,
    defaultCurrency: DEFAULT_CURRENCY,
  },
  exchangeRates: {
    rates: {
      EUR: 1,
      CHF: 1.0784,
      HKD: 9.0882,
      SEK: 10.2958,
      USD: 1.1726,
      GBP: 0.90013,
      HUF: 345.72,
      CZK: 26.319,
      AUD: 1.6508,
      HRK: 7.4755,
      PLN: 4.4201,
    },
    base: 'EUR',
    date: '2020-08-03',
  },
})
