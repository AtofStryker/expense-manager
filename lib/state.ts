import {
  createDefaultAddTransactionState,
  AddTransaction,
  Tag,
  Transaction
} from './addTransaction/state'
import { ObjectOf } from './types'

export type ScreenTitle =
  | 'add'
  | 'transactions'
  | 'charts'
  | 'tags'
  | 'settings'

export type SignInStatus = 'loggedIn' | 'loggingIn' | 'loggedOut' | 'unknown'

export interface State {
  cnt: number
  // use firebase.auth().currentUser to get the current user
  signInStatus: SignInStatus
  currentScreen: ScreenTitle
  addTransaction: AddTransaction
  tags: ObjectOf<Tag>
  transactions: ObjectOf<Transaction>
}

const state: State = {
  cnt: 0,
  signInStatus: 'unknown',
  currentScreen: 'add',
  addTransaction: createDefaultAddTransactionState([]),
  tags: {},
  transactions: {},
}

export const getInitialState = () => state
