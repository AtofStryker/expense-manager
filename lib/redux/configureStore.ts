import { applyMiddleware, createStore } from 'redux'
import { createLogger } from 'redux-logger'
import thunk from 'redux-thunk'

import { Api } from '../api'
import { getInitialState } from '../state'
import { Logger } from '../types'

import rootReducer from './rootReducer'
import { Action, ThunkExtraArgument } from './types'

export const configureStore = () => {
  const logger: Logger = {
    log: (_, __) => null,
  }
  if (process.env.NODE_ENV === 'development') {
    logger.log = (message, payload) =>
      store.dispatch({
        type: message,
        payload,
      } as Action<any>)
  }

  const loggerMiddleware = createLogger({
    collapsed: true,
    predicate: (_, action: Action) => !(action.loggable === false),
    actionTransformer: (action: Action) => ({
      ...action,
      type: `${action.type}`,
    }),
  })

  const thunkExtra: ThunkExtraArgument = {
    logger,
    api: new Api(logger),
  }
  const middlewares = [thunk.withExtraArgument(thunkExtra)]
  if (process.env.NODE_ENV === 'development') {
    middlewares.push(loggerMiddleware)
  }

  const store = createStore(
    rootReducer as any,
    getInitialState() as any,
    applyMiddleware(...middlewares),
  )

  return store
}
