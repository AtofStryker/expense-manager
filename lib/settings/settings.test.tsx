import { mount } from 'enzyme'
import { Store } from 'redux'

import { getInitialState, State } from '../state'
import {
  byAriaLabel,
  configureTestStore,
  getMockedFirebase,
  initializeMockFirebase,
  reduxify
} from '../testing'

import Settings from './index'

jest.mock('../firebase/firebase', () => initializeMockFirebase())

describe('settings', () => {
  test('shows loading page', () => {
    const comp = mount(
      reduxify(Settings, configureTestStore(getInitialState())),
    )
    expect(comp.html()).toMatchSnapshot()
  })

  describe('logged in', () => {
    let state: State
    let store: Store
    let firebase: ReturnType<typeof getMockedFirebase>

    beforeEach(() => {
      state = getInitialState()
      firebase = getMockedFirebase()
      store = configureTestStore(state)
      state.signInStatus = 'loggedIn'
      firebase.auth().currentUser = {
        uid: 'userId',
      } as firebase.User
    })

    test('shows settings page', () => {
      const comp = mount(reduxify(Settings, store))
      expect(comp.html()).toMatchSnapshot()
    })

    test('signs out', () => {
      const comp = mount(reduxify(Settings, store))
      expect(firebase.auth().currentUser).not.toBe(null)
      comp
        .find(byAriaLabel('sign out'))
        .hostNodes()
        .simulate('click')
      expect(firebase.auth().currentUser).toBe(null)
    })
  })
})
