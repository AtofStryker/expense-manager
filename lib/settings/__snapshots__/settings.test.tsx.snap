// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`settings logged in shows settings page 1`] = `"<div class=\\"MuiBottomNavigation-root makeStyles-bottomNav-33\\"><button class=\\"MuiButtonBase-root MuiBottomNavigationAction-root\\" tabindex=\\"0\\" type=\\"button\\" style=\\"text-transform: capitalize; min-width: unset; padding: unset;\\"><span class=\\"MuiBottomNavigationAction-wrapper\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z\\"></path></svg><span class=\\"MuiBottomNavigationAction-label\\">add</span></span><span class=\\"MuiTouchRipple-root\\"></span></button><button class=\\"MuiButtonBase-root MuiBottomNavigationAction-root\\" tabindex=\\"0\\" type=\\"button\\" style=\\"text-transform: capitalize; min-width: unset; padding: unset;\\"><span class=\\"MuiBottomNavigationAction-wrapper\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M3 13h2v-2H3v2zm0 4h2v-2H3v2zm0-8h2V7H3v2zm4 4h14v-2H7v2zm0 4h14v-2H7v2zM7 7v2h14V7H7z\\"></path></svg><span class=\\"MuiBottomNavigationAction-label\\">transactions</span></span><span class=\\"MuiTouchRipple-root\\"></span></button><button class=\\"MuiButtonBase-root MuiBottomNavigationAction-root\\" tabindex=\\"0\\" type=\\"button\\" style=\\"text-transform: capitalize; min-width: unset; padding: unset;\\"><span class=\\"MuiBottomNavigationAction-wrapper\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M5 9.2h3V19H5zM10.6 5h2.8v14h-2.8zm5.6 8H19v6h-2.8z\\"></path></svg><span class=\\"MuiBottomNavigationAction-label\\">charts</span></span><span class=\\"MuiTouchRipple-root\\"></span></button><button class=\\"MuiButtonBase-root MuiBottomNavigationAction-root\\" tabindex=\\"0\\" type=\\"button\\" style=\\"text-transform: capitalize; min-width: unset; padding: unset;\\"><span class=\\"MuiBottomNavigationAction-wrapper\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M21.41 11.58l-9-9C12.05 2.22 11.55 2 11 2H4c-1.1 0-2 .9-2 2v7c0 .55.22 1.05.59 1.42l9 9c.36.36.86.58 1.41.58.55 0 1.05-.22 1.41-.59l7-7c.37-.36.59-.86.59-1.41 0-.55-.23-1.06-.59-1.42zM5.5 7C4.67 7 4 6.33 4 5.5S4.67 4 5.5 4 7 4.67 7 5.5 6.33 7 5.5 7z\\"></path></svg><span class=\\"MuiBottomNavigationAction-label\\">tags</span></span><span class=\\"MuiTouchRipple-root\\"></span></button><button class=\\"MuiButtonBase-root MuiBottomNavigationAction-root Mui-selected\\" tabindex=\\"0\\" type=\\"button\\" style=\\"text-transform: capitalize; min-width: unset; padding: unset;\\"><span class=\\"MuiBottomNavigationAction-wrapper\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path transform=\\"scale(1.2, 1.2)\\" fill=\\"none\\" d=\\"M0 0h20v20H0V0z\\"></path><path transform=\\"scale(1.2, 1.2)\\" d=\\"M15.95 10.78c.03-.25.05-.51.05-.78s-.02-.53-.06-.78l1.69-1.32c.15-.12.19-.34.1-.51l-1.6-2.77c-.1-.18-.31-.24-.49-.18l-1.99.8c-.42-.32-.86-.58-1.35-.78L12 2.34c-.03-.2-.2-.34-.4-.34H8.4c-.2 0-.36.14-.39.34l-.3 2.12c-.49.2-.94.47-1.35.78l-1.99-.8c-.18-.07-.39 0-.49.18l-1.6 2.77c-.1.18-.06.39.1.51l1.69 1.32c-.04.25-.07.52-.07.78s.02.53.06.78L2.37 12.1c-.15.12-.19.34-.1.51l1.6 2.77c.1.18.31.24.49.18l1.99-.8c.42.32.86.58 1.35.78l.3 2.12c.04.2.2.34.4.34h3.2c.2 0 .37-.14.39-.34l.3-2.12c.49-.2.94-.47 1.35-.78l1.99.8c.18.07.39 0 .49-.18l1.6-2.77c.1-.18.06-.39-.1-.51l-1.67-1.32zM10 13c-1.65 0-3-1.35-3-3s1.35-3 3-3 3 1.35 3 3-1.35 3-3 3z\\"></path></svg><span class=\\"MuiBottomNavigationAction-label Mui-selected\\">settings</span></span><span class=\\"MuiTouchRipple-root\\"></span></button></div><div class=\\"makeStyles-root-1\\">settings<button aria-label=\\"sign out\\">google sign out</button><div class=\\"MuiPaper-root MuiPaper-elevation1 MuiExpansionPanel-root MuiExpansionPanel-rounded MuiPaper-rounded\\" aria-label=\\"import and export\\" style=\\"text-transform: capitalize;\\"><div class=\\"MuiButtonBase-root MuiExpansionPanelSummary-root\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\" aria-expanded=\\"false\\"><div class=\\"MuiExpansionPanelSummary-content\\"><p class=\\"MuiTypography-root MuiTypography-body1\\">import and export</p></div><div class=\\"MuiButtonBase-root MuiIconButton-root MuiExpansionPanelSummary-expandIcon MuiIconButton-edgeEnd\\" tabindex=\\"-1\\" role=\\"button\\" aria-disabled=\\"false\\" aria-hidden=\\"true\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z\\"></path></svg></span><span class=\\"MuiTouchRipple-root\\"></span></div></div><div class=\\"MuiCollapse-container MuiCollapse-hidden\\" style=\\"min-height: 0px;\\"><div class=\\"MuiCollapse-wrapper\\"><div class=\\"MuiCollapse-wrapperInner\\"><div role=\\"region\\"><div class=\\"MuiExpansionPanelDetails-root makeStyles-importExportContent-52\\"><input id=\\"choose-import-file\\" type=\\"file\\" style=\\"display: none;\\"><label for=\\"choose-import-file\\"><span class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-2 MuiButton-containedPrimary MuiButton-fullWidth\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\" aria-label=\\"import from csv file\\"><span class=\\"MuiButton-label\\">Import from csv file</span><span class=\\"MuiTouchRipple-root\\"></span></span></label><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-2 MuiButton-containedPrimary MuiButton-fullWidth\\" tabindex=\\"0\\" type=\\"button\\" aria-label=\\"export to csv file\\"><span class=\\"MuiButton-label\\">Export to csv file</span><span class=\\"MuiTouchRipple-root\\"></span></button></div></div></div></div></div></div><div class=\\"MuiPaper-root MuiPaper-elevation1 MuiExpansionPanel-root MuiExpansionPanel-rounded MuiPaper-rounded\\" aria-label=\\"clear data\\" style=\\"text-transform: capitalize;\\"><div class=\\"MuiButtonBase-root MuiExpansionPanelSummary-root\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\" aria-expanded=\\"false\\"><div class=\\"MuiExpansionPanelSummary-content\\"><p class=\\"MuiTypography-root MuiTypography-body1\\">clear data</p></div><div class=\\"MuiButtonBase-root MuiIconButton-root MuiExpansionPanelSummary-expandIcon MuiIconButton-edgeEnd\\" tabindex=\\"-1\\" role=\\"button\\" aria-disabled=\\"false\\" aria-hidden=\\"true\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z\\"></path></svg></span><span class=\\"MuiTouchRipple-root\\"></span></div></div><div class=\\"MuiCollapse-container MuiCollapse-hidden\\" style=\\"min-height: 0px;\\"><div class=\\"MuiCollapse-wrapper\\"><div class=\\"MuiCollapse-wrapperInner\\"><div role=\\"region\\"><div class=\\"MuiExpansionPanelDetails-root makeStyles-importExportContent-52\\"><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-2 MuiButton-containedSecondary MuiButton-fullWidth\\" tabindex=\\"0\\" type=\\"button\\" aria-label=\\"clear data\\"><span class=\\"MuiButton-label\\">Clear all expense manager data</span><span class=\\"MuiTouchRipple-root\\"></span></button></div></div></div></div></div></div></div>"`;

exports[`settings shows loading page 1`] = `
"<img src=\\"../static/coin.svg\\" alt=\\"coin\\" style=\\"width: 60vw; margin-top: 20vh;\\"><h3 class=\\"MuiTypography-root MuiTypography-h3 MuiTypography-gutterBottom\\" style=\\"text-align: center; margin-top: 10vh;\\">Loading...</h3><style>
        @keyframes rotation {
          from {
            transform: rotateY(0deg);
          }
          to {
            transform: rotateY(1800deg);
          }
        }

        img {
          animation: rotation 2s infinite ease-in-out;
          display: block;
          margin-left: auto;
          margin-right: auto;
        }
      </style>"
`;
